File: src/tickets.js
const {
  ChannelType, PermissionFlagsBits, EmbedBuilder, ActionRowBuilder, ButtonBuilder,
  ButtonStyle, AttachmentBuilder, ModalBuilder, TextInputBuilder, TextInputStyle,
  StringSelectMenuBuilder
} = require('discord.js');
const database = require('./database');
const { COLORS, truncate } = require('./utils/ui');
const { sendModLog } = require('./utils/modLog');

function ticketConfig(guildId) {
  return {
    supportRoleId: database.getConfig(guildId, 'ticket_support_role_id'),
    panelTitle: database.getConfig(guildId, 'ticket_panel_title', '🎫 Destek Talebi'),
    panelDescription: database.getConfig(guildId, 'ticket_panel_description', 'Destek ekibimize ulaşmak için aşağıdaki düğmeye bas. Aynı anda yalnızca bir açık talebin olabilir.'),
    welcomeMessage: database.getConfig(guildId, 'ticket_welcome_message', '{user}, talebin oluşturuldu. Sorununu ayrıntılı biçimde anlat; ekibimiz kısa sürede ilgilenecek.'),
    buttonLabel: database.getConfig(guildId, 'ticket_panel_button_label', 'Destek Talebi Aç'),
    buttonEmoji: database.getConfig(guildId, 'ticket_panel_button_emoji', '🎫')
  };
}

function buildTicketPanel(guild) {
  const config = guild ? ticketConfig(guild.id) : ticketConfig('0');
  return {
    embeds: [new EmbedBuilder().setColor(COLORS.primary).setTitle(truncate(config.panelTitle, 256)).setDescription(truncate(config.panelDescription, 4000))
      .setFooter({ text: 'Bir kullanıcı aynı anda yalnızca bir açık talep oluşturabilir.' })],
    components: [new ActionRowBuilder().addComponents(
      new ButtonBuilder().setCustomId('ticket_open').setLabel(truncate(config.buttonLabel, 80)).setEmoji(config.buttonEmoji || '🎫').setStyle(ButtonStyle.Primary)
    )]
  };
}

function ticketOpenModal() {
  const modal = new ModalBuilder().setCustomId('ticket_open_modal').setTitle('Destek Talebi Oluştur');
  modal.addComponents(
    new ActionRowBuilder().addComponents(new TextInputBuilder().setCustomId('subject').setLabel('Konu').setStyle(TextInputStyle.Short).setRequired(true).setMaxLength(100)),
    new ActionRowBuilder().addComponents(new TextInputBuilder().setCustomId('description').setLabel('Sorunun / talebin').setStyle(TextInputStyle.Paragraph).setRequired(true).setMinLength(10).setMaxLength(1500)),
    new ActionRowBuilder().addComponents(new TextInputBuilder().setCustomId('priority').setLabel('Öncelik: düşük / normal / yüksek / acil').setStyle(TextInputStyle.Short).setRequired(true).setValue('normal').setMaxLength(10))
  );
  return modal;
}

function ticketAddModal() {
  const modal = new ModalBuilder().setCustomId('ticket_add_modal').setTitle('Kanalına Üye Ekle');
  modal.addComponents(
    new ActionRowBuilder().addComponents(new TextInputBuilder().setCustomId('user_input').setLabel('Eklenecek Üye ID veya Tag').setStyle(TextInputStyle.Short).setRequired(true).setPlaceholder('Örn: 123456789012345678'))
  );
  return modal;
}

function ticketRenameModal() {
  const modal = new ModalBuilder().setCustomId('ticket_rename_modal').setTitle('Ticket Kanalını Adlandır');
  modal.addComponents(
    new ActionRowBuilder().addComponents(new TextInputBuilder().setCustomId('name_input').setLabel('Yeni Kanal Adı').setStyle(TextInputStyle.Short).setRequired(true).setMaxLength(80))
  );
  return modal;
}

function ticketCloseModal() {
  const modal = new ModalBuilder().setCustomId('ticket_close_modal').setTitle('Ticket Talebini Kapat');
  modal.addComponents(
    new ActionRowBuilder().addComponents(new TextInputBuilder().setCustomId('reason_input').setLabel('Kapanış Sebebi / Notu').setStyle(TextInputStyle.Paragraph).setRequired(false).setValue('Çözüldü / Talebiniz tamamlandı.').setMaxLength(500))
  );
  return modal;
}

function ticketPermissions(guild, ownerId, supportRoleId) {
  const permissions = [
    { id: guild.id, deny: [PermissionFlagsBits.ViewChannel] },
    { id: ownerId, allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages, PermissionFlagsBits.ReadMessageHistory, PermissionFlagsBits.AttachFiles] },
    { id: guild.members.me.id, allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages, PermissionFlagsBits.ReadMessageHistory, PermissionFlagsBits.ManageChannels, PermissionFlagsBits.ManageMessages] }
  ];
  if (supportRoleId) permissions.push({ id: supportRoleId, allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages, PermissionFlagsBits.ReadMessageHistory] });
  return permissions;
}

function isSupport(member, guildId) {
  const supportRoleId = database.getConfig(guildId, 'ticket_support_role_id');
  return member.permissions.has(PermissionFlagsBits.ManageChannels) || Boolean(supportRoleId && member.roles.cache.has(supportRoleId));
}

async function ticketLogChannel(guild) {
  const channelId = database.getSettings(guild.id).ticket_log_channel_id;
  return channelId ? guild.channels.fetch(channelId).catch(() => null) : null;
}

async function sendTicketLog(guild, payload) {
  const channel = await ticketLogChannel(guild);
  if (channel?.isTextBased()) await channel.send(payload).catch(() => null);
}

async function createTranscript(channel) {
  const messages = await channel.messages.fetch({ limit: 100 }).catch(() => null);
  if (!messages) return null;
  const lines = [...messages.values()].reverse().map((message) => {
    const date = new Date(message.createdTimestamp).toISOString();
    const attachments = [...message.attachments.values()].map((item) => item.url).join(' ');
    return `[${date}] ${message.author.tag} (${message.author.id}): ${message.content || '[embed/boş mesaj]'} ${attachments}`.trim();
  });
  return new AttachmentBuilder(Buffer.from(lines.join('\n'), 'utf8'), { name: `ticket-${channel.id}.txt` });
}

async function openTicket(interaction) {
  const guild = interaction.guild;
  const settings = database.getSettings(guild.id);
  const config = ticketConfig(guild.id);
  const existing = database.getOpenTicket(guild.id, interaction.user.id);
  if (existing) return interaction.reply({ content: `zaten açık bir talebin var: <#${existing.channel_id}>`, flags: 64 });
  if (!guild.members.me.permissions.has(PermissionFlagsBits.ManageChannels)) return interaction.reply({ content: 'botta `Kanalları Yönet` yetkisi yok.', flags: 64 });
  const category = settings.ticket_category_id ? await guild.channels.fetch(settings.ticket_category_id).catch(() => null) : null;
  if (settings.ticket_category_id && category?.type !== ChannelType.GuildCategory && category?.type !== 4) {
    const embed = new EmbedBuilder().setColor(COLORS.danger).setTitle('❌ Geçersiz Ticket Kategorisi')
      .setDescription('Sunucuda ayarlı olan Ticket Kategorisi bir metin kanalı olarak kaydedilmiş. Yetkililer `/ticketsetup` veya `/kurulum` ile bir **Discord Kategori Başlığı (📁 GuildCategory)** seçmelidir.');
    return interaction.reply({ embeds: [embed], flags: 64 });
  }
  const details = {
    type: 'destek',
    priority: interaction.fields.getTextInputValue('priority').toLocaleLowerCase('tr-TR'),
    subject: interaction.fields.getTextInputValue('subject'),
    description: interaction.fields.getTextInputValue('description')
  };
  if (!['düşük', 'dusuk', 'normal', 'yüksek', 'yuksek', 'acil'].includes(details.priority)) {
    return interaction.reply({ content: 'öncelik yalnızca düşük, normal, yüksek veya acil olabilir.', flags: 64 });
  }
  const priority = details.priority === 'dusuk' ? 'düşük' : details.priority === 'yuksek' ? 'yüksek' : details.priority;
  const safeName = `${priority}-${interaction.user.username}`.toLocaleLowerCase('tr-TR').replace(/[^a-z0-9çğıöşü-]/g, '-').slice(0, 70) || interaction.user.id;
  const ticket = await guild.channels.create({
    name: `talep-${safeName}`,
    type: ChannelType.GuildText,
    parent: category?.id,
    permissionOverwrites: ticketPermissions(guild, interaction.user.id, config.supportRoleId),
    topic: `Ticket sahibi: ${interaction.user.tag} (${interaction.user.id}) · ${details.subject}`,
    reason: `Ticket: ${interaction.user.tag}`
  });
  database.createTicket(ticket.id, guild.id, interaction.user.id, category?.id, { ...details, priority });
  await sendTicketLog(guild, { content: `🎫 Ticket açıldı: ${ticket} · sahibi: <@${interaction.user.id}> · öncelik: **${priority}** · konu: **${truncate(details.subject, 100)}**` });
  const welcome = config.welcomeMessage.replaceAll('{user}', `<@${interaction.user.id}>`).replaceAll('{username}', interaction.user.username);
  const embed = new EmbedBuilder().setColor(priority === 'acil' ? COLORS.danger : priority === 'yüksek' ? COLORS.warning : COLORS.success)
    .setTitle(`🎫 ${details.subject}`).setDescription(`${truncate(welcome, 2500)}\n\n**Açıklama**\n${truncate(details.description, 1200)}`)
    .addFields(
      { name: 'Talep sahibi', value: `<@${interaction.user.id}>`, inline: true },
      { name: 'Öncelik', value: priority, inline: true },
      { name: 'Durum', value: '🟢 Açık', inline: true }
    ).setTimestamp();
  const row1 = new ActionRowBuilder().addComponents(
    new ButtonBuilder().setCustomId('ticket_claim').setLabel('Talebi Üstlen').setEmoji('🙋').setStyle(ButtonStyle.Success),
    new ButtonBuilder().setCustomId('ticket_add_btn').setLabel('Üye Ekle').setEmoji('➕').setStyle(ButtonStyle.Primary),
    new ButtonBuilder().setCustomId('ticket_rename_btn').setLabel('Adlandır').setEmoji('✏️').setStyle(ButtonStyle.Secondary)
  );
  const row2 = new ActionRowBuilder().addComponents(
    new ButtonBuilder().setCustomId('ticket_status').setLabel('Durum Değiştir').setEmoji('📌').setStyle(ButtonStyle.Primary),
    new ButtonBuilder().setCustomId('ticket_close_btn').setLabel('Talebi Kapat').setEmoji('🔒').setStyle(ButtonStyle.Danger)
  );
  await ticket.send({
    content: config.supportRoleId ? `<@&${config.supportRoleId}>` : undefined,
    embeds: [embed],
    components: [row1, row2],
    allowedMentions: { roles: config.supportRoleId ? [config.supportRoleId] : [] }
  });
  await interaction.reply({ content: `talebin oluşturuldu: ${ticket}`, flags: 64 });
}

async function claimTicket(interaction) {
  const ticket = database.getTicket(interaction.channelId);
  if (!ticket || ticket.closed_at) return interaction.followUp({ content: 'bu kanal açık bir ticket değil.', flags: 64 });
  if (!isSupport(interaction.member, interaction.guildId)) return interaction.followUp({ content: 'ticket destek rolüne veya `Kanalları Yönet` yetkisine sahip değilsin.', flags: 64 });
  if (!database.claimTicket(interaction.channelId, interaction.user.id)) return interaction.followUp({ content: ticket.claimed_by_id ? `talep zaten <@${ticket.claimed_by_id}> tarafından üstlenildi.` : 'talep üstlenilemedi.', flags: 64 });
  await interaction.channel.send(`🙋 Bu talebi <@${interaction.user.id}> üstlendi.`);
  await sendTicketLog(interaction.guild, { content: `🙋 Ticket üstlenildi: <#${ticket.channel_id}> · yetkili: <@${interaction.user.id}>` });
}

async function closeTicket(interaction, customReason) {
  const ticket = database.getTicket(interaction.channelId);
  const reason = customReason || (interaction.fields ? interaction.fields.getTextInputValue('reason_input') : 'Sebep belirtilmedi');
  const respond = interaction.replied || interaction.deferred
    ? (payload) => interaction.followUp(payload)
    : (payload) => interaction.reply(payload);

  if (!ticket || ticket.closed_at) return respond({ content: 'bu kanal açık bir ticket değil.', flags: 64 });
  if (ticket.owner_id !== interaction.user.id && !isSupport(interaction.member, interaction.guildId)) return respond({ content: 'bu talebi kapatma yetkin yok.', flags: 64 });
  if (!database.closeTicket(interaction.channelId, interaction.user.id, reason)) return respond({ content: 'ticket zaten kapatılmış.', flags: 64 });

  const transcript = await createTranscript(interaction.channel);
  await sendModLog(interaction.guild, { action: 'ticket', targetId: ticket.owner_id, moderatorId: interaction.user.id, reason, color: COLORS.neutral });
  await sendTicketLog(interaction.guild, {
    content: `🔒 Ticket kapatıldı: **${interaction.channel.name}** · sahibi: <@${ticket.owner_id}> · kapatan: <@${interaction.user.id}> · sebep: **${truncate(reason, 200)}**`,
    files: transcript ? [transcript] : []
  });

  const embed = new EmbedBuilder().setColor(COLORS.danger).setTitle('🔒 Ticket Kapatıldı')
    .setDescription(`Bu destek talebi **<@${interaction.user.id}>** tarafından kapatıldı.\n\n**Kapanış Sebebi:**\n${truncate(reason, 500)}\n\n*Transcript log kanalına gönderildi. Kanal 5 saniye içinde silinecektir.*`)
    .setTimestamp();

  await interaction.channel.send({ embeds: [embed] }).catch(() => null);
  setTimeout(() => interaction.channel.delete(`Ticket kapatıldı: ${interaction.user.tag}`).catch(() => null), 5000).unref();
}

module.exports = {
  buildTicketPanel,
  ticketOpenModal,
  ticketAddModal,
  ticketRenameModal,
  ticketCloseModal,
  openTicket,
  claimTicket,
  closeTicket,
  isSupport,
  ticketConfig
};

File: internal/bot/tickets.go
package bot

import (
	"bytes"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/adashlabs/adash-bot/internal/database"
	"github.com/bwmarrin/discordgo"
)

func (b *Bot) ticketPanel(guild string) (*discordgo.MessageEmbed, []discordgo.MessageComponent) {
	title := b.db.ConfigString(guild, "ticket_panel_title", "🎫 Destek Talebi")
	desc := b.db.ConfigString(guild, "ticket_panel_description", "Destek ekibimize ulaşmak için aşağıdaki düğmeye bas. Aynı anda yalnızca bir açık talebin olabilir.")
	label := b.db.ConfigString(guild, "ticket_panel_button_label", "Destek Talebi Aç")
	emoji := b.db.ConfigString(guild, "ticket_panel_button_emoji", "🎫")
	em := embed(trunc(title, 256), trunc(desc, 4000), colorPrimary)
	em.Footer = &discordgo.MessageEmbedFooter{Text: "Bir kullanıcı aynı anda yalnızca bir açık talep oluşturabilir."}
	return em, []discordgo.MessageComponent{row(button("ticket_open", trunc(label, 80), discordgo.PrimaryButton, emoji))}
}
func (b *Bot) sendTicketPanel(guild, channel string) error {
	em, components := b.ticketPanel(guild)
	_, e := b.dg.ChannelMessageSendComplex(channel, &discordgo.MessageSend{Embeds: []*discordgo.MessageEmbed{em}, Components: components})
	return e
}
func (b *Bot) ticketSetupCommand(c *commandContext, args []string) error {
	if e := c.require(discordgo.PermissionManageServer); e != nil {
		return e
	}
	if len(args) < 2 {
		em, components := b.ticketSetupWizard(c.guildID)
		return c.embed(em, components...)
	}
	category, panel := mentionID(args[0]), mentionID(args[1])
	cat, e := c.s.Channel(category)
	if e != nil || cat.Type != discordgo.ChannelTypeGuildCategory {
		return fmt.Errorf("geçerli bir kategori belirt")
	}
	ch, e := c.s.Channel(panel)
	if e != nil || ch.Type != discordgo.ChannelTypeGuildText {
		return fmt.Errorf("geçerli panel kanalı belirt")
	}
	if e = b.db.SetSetting(c.guildID, "ticket_category_id", category); e != nil {
		return e
	}
	if len(args) > 2 {
		if id := mentionID(args[2]); id != "" {
			_ = b.db.SetSetting(c.guildID, "ticket_log_channel_id", id)
		}
	}
	if len(args) > 3 {
		if id := mentionID(args[3]); id != "" {
			_ = b.db.SetConfig(c.guildID, "ticket_support_role_id", id)
		}
	}
	if e = b.sendTicketPanel(c.guildID, panel); e != nil {
		return e
	}
	return c.embed(successEmbed("🎫 Ticket Sistemi Kuruldu", fmt.Sprintf("Kategori: <#%s>\nPanel: <#%s>", category, panel)))
}
func (b *Bot) isSupport(c *commandContext) bool {
	p, e := c.permissions()
	if e == nil && (p&discordgo.PermissionManageChannels != 0 || p&discordgo.PermissionAdministrator != 0) {
		return true
	}
	role := b.db.ConfigString(c.guildID, "ticket_support_role_id", "")
	if role != "" && c.member != nil {
		for _, r := range c.member.Roles {
			if r == role {
				return true
			}
		}
	}
	return false
}
func (b *Bot) ticketCommand(c *commandContext, args []string) error {
	t, e := b.db.Ticket(c.channelID)
	if e != nil || t.ClosedAt.Valid {
		return fmt.Errorf("bu komut yalnızca açık ticket kanalında kullanılabilir")
	}
	if !b.isSupport(c) {
		return fmt.Errorf("ticket destek rolü veya Kanalları Yönet yetkisi gerekiyor")
	}
	if len(args) == 0 {
		return c.embed(embed("🎫 Ticket Yönetimi", "`ticket ekle @üye`\n`ticket çıkar @üye`\n`ticket adlandır yeni-ad`", colorPrimary))
	}
	action := strings.ToLower(args[0])
	if action == "ekle" || action == "add" || action == "çıkar" || action == "cikar" || action == "remove" {
		if len(args) < 2 {
			return fmt.Errorf("kullanıcı belirt")
		}
		u, m, e := b.target(c, args[1])
		if e != nil || m == nil {
			return fmt.Errorf("sunucudaki kullanıcı bulunamadı")
		}
		if action == "çıkar" || action == "cikar" || action == "remove" {
			if u.ID == t.OwnerID {
				return fmt.Errorf("ticket sahibi çıkarılamaz")
			}
			if e = c.s.ChannelPermissionDelete(c.channelID, u.ID); e != nil {
				return e
			}
			return c.text("👤 <@" + u.ID + "> ticket kanalından çıkarıldı.")
		}
		allow := int64(discordgo.PermissionViewChannel | discordgo.PermissionSendMessages | discordgo.PermissionReadMessageHistory)
		if e = c.s.ChannelPermissionSet(c.channelID, u.ID, discordgo.PermissionOverwriteTypeMember, allow, 0); e != nil {
			return e
		}
		return c.text("👤 <@" + u.ID + "> ticket kanalına eklendi.")
	}
	if action == "adlandır" || action == "adlandir" || action == "rename" {
		name := safeChannelName(strings.Join(args[1:], "-"))
		if len(name) < 2 {
			return fmt.Errorf("geçerli kanal adı belirt")
		}
		if _, e = c.s.ChannelEdit(c.channelID, &discordgo.ChannelEdit{Name: name}); e != nil {
			return e
		}
		return c.text("✏️ Ticket kanalının adı **" + name + "** oldu.")
	}
	return fmt.Errorf("geçersiz ticket işlemi")
}
func safeChannelName(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	s = regexp.MustCompile(`[^a-z0-9çğıöşü-]+`).ReplaceAllString(s, "-")
	s = regexp.MustCompile(`-+`).ReplaceAllString(s, "-")
	return trunc(strings.Trim(s, "-"), 80)
}
func (b *Bot) ticketOpenModal(s *discordgo.Session, i *discordgo.InteractionCreate) error {
	inputs := []discordgo.MessageComponent{row(discordgo.TextInput{CustomID: "subject", Label: "Konu", Style: discordgo.TextInputShort, Required: true, MaxLength: 100}), row(discordgo.TextInput{CustomID: "description", Label: "Sorunun / talebin", Style: discordgo.TextInputParagraph, Required: true, MinLength: 10, MaxLength: 1500}), row(discordgo.TextInput{CustomID: "priority", Label: "Öncelik: düşük / normal / yüksek / acil", Style: discordgo.TextInputShort, Required: true, Value: "normal", MaxLength: 10})}
	return s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{Type: discordgo.InteractionResponseModal, Data: &discordgo.InteractionResponseData{CustomID: "ticket_open_modal", Title: "Destek Talebi Oluştur", Components: inputs}})
}
func (b *Bot) openTicketFromModal(s *discordgo.Session, i *discordgo.InteractionCreate, v map[string]string) error {
	existing, e := b.db.OpenTicket(i.GuildID, userOf(i).ID)
	if e == nil {
		return ephemeral(s, i, "Zaten açık bir talebin var: <#"+existing.ChannelID+">")
	}
	priority := strings.ToLower(v["priority"])
	if priority == "dusuk" {
		priority = "düşük"
	}
	if priority == "yuksek" {
		priority = "yüksek"
	}
	valid := map[string]bool{"düşük": true, "normal": true, "yüksek": true, "acil": true}
	if !valid[priority] {
		return fmt.Errorf("öncelik düşük, normal, yüksek veya acil olmalı")
	}
	settings, e := b.db.Settings(i.GuildID)
	if e != nil {
		return e
	}
	category := ""
	if settings.TicketCategoryID.Valid {
		category = settings.TicketCategoryID.String
	}
	overwrites := []*discordgo.PermissionOverwrite{{ID: i.GuildID, Type: discordgo.PermissionOverwriteTypeRole, Deny: discordgo.PermissionViewChannel}, {ID: userOf(i).ID, Type: discordgo.PermissionOverwriteTypeMember, Allow: discordgo.PermissionViewChannel | discordgo.PermissionSendMessages | discordgo.PermissionReadMessageHistory | discordgo.PermissionAttachFiles}, {ID: s.State.User.ID, Type: discordgo.PermissionOverwriteTypeMember, Allow: discordgo.PermissionViewChannel | discordgo.PermissionSendMessages | discordgo.PermissionReadMessageHistory | discordgo.PermissionManageChannels | discordgo.PermissionManageMessages}}
	support := b.db.ConfigString(i.GuildID, "ticket_support_role_id", "")
	if support != "" {
		overwrites = append(overwrites, &discordgo.PermissionOverwrite{ID: support, Type: discordgo.PermissionOverwriteTypeRole, Allow: discordgo.PermissionViewChannel | discordgo.PermissionSendMessages | discordgo.PermissionReadMessageHistory})
	}
	name := "talep-" + safeChannelName(priority+"-"+userOf(i).Username)
	ch, e := s.GuildChannelCreateComplex(i.GuildID, discordgo.GuildChannelCreateData{Name: name, Type: discordgo.ChannelTypeGuildText, ParentID: category, Topic: fmt.Sprintf("Ticket sahibi: %s (%s) · %s", userOf(i).Username, userOf(i).ID, v["subject"]), PermissionOverwrites: overwrites})
	if e != nil {
		return e
	}
	if e = b.db.CreateTicket(ch.ID, i.GuildID, userOf(i).ID, category, "destek", priority, v["subject"], v["description"]); e != nil {
		_, _ = s.ChannelDelete(ch.ID)
		return e
	}
	welcome := b.db.ConfigString(i.GuildID, "ticket_welcome_message", "{user}, talebin oluşturuldu. Sorununu ayrıntılı biçimde anlat.")
	welcome = strings.ReplaceAll(strings.ReplaceAll(welcome, "{user}", "<@"+userOf(i).ID+">"), "{username}", userOf(i).Username)
	em := embed("🎫 "+trunc(v["subject"], 240), trunc(welcome, 2500)+"\n\n**Açıklama**\n"+trunc(v["description"], 1200), strColor(priority == "acil", colorDanger, strColor(priority == "yüksek", colorWarning, colorSuccess)))
	em.Fields = []*discordgo.MessageEmbedField{{Name: "Talep sahibi", Value: "<@" + userOf(i).ID + ">", Inline: true}, {Name: "Öncelik", Value: priority, Inline: true}, {Name: "Durum", Value: "🟢 Açık", Inline: true}}
	components := ticketControls()
	content := ""
	if support != "" {
		content = "<@&" + support + ">"
	}
	_, _ = s.ChannelMessageSendComplex(ch.ID, &discordgo.MessageSend{Content: content, Embeds: []*discordgo.MessageEmbed{em}, Components: components, AllowedMentions: &discordgo.MessageAllowedMentions{Roles: []string{support}}})
	b.ticketLog(i.GuildID, "🎫 Ticket açıldı: <#"+ch.ID+"> · sahibi: <@"+userOf(i).ID+">", nil)
	return ephemeral(s, i, "Talebin oluşturuldu: <#"+ch.ID+">")
}
func ticketControls() []discordgo.MessageComponent {
	return []discordgo.MessageComponent{row(button("ticket_claim", "Talebi Üstlen", discordgo.SuccessButton, "🙋"), button("ticket_add_btn", "Üye Ekle", discordgo.PrimaryButton, "➕"), button("ticket_rename_btn", "Adlandır", discordgo.SecondaryButton, "✏️")), row(button("ticket_status", "Durum Değiştir", discordgo.PrimaryButton, "📌"), button("ticket_close_btn", "Talebi Kapat", discordgo.DangerButton, "🔒"))}
}
func (b *Bot) ticketComponent(s *discordgo.Session, i *discordgo.InteractionCreate) error {
	id := i.MessageComponentData().CustomID
	c := &commandContext{b: b, s: s, guildID: i.GuildID, channelID: i.ChannelID, user: userOf(i), member: i.Member}
	t, e := b.db.Ticket(i.ChannelID)
	if e != nil || t.ClosedAt.Valid {
		return fmt.Errorf("bu kanal açık ticket değil")
	}
	if id == "ticket_close_btn" {
		return ticketModal(s, i, "ticket_close_modal", "Ticket Talebini Kapat", "reason_input", "Kapanış Sebebi / Notu", discordgo.TextInputParagraph, "Çözüldü / Talebiniz tamamlandı.")
	}
	if !b.isSupport(c) {
		return fmt.Errorf("ticket destek rolü veya Kanalları Yönet yetkisi gerekiyor")
	}
	switch id {
	case "ticket_claim":
		ok, e := b.db.ClaimTicket(i.ChannelID, userOf(i).ID)
		if e != nil {
			return e
		}
		if !ok {
			return fmt.Errorf("talep zaten üstlenildi")
		}
		_ = ephemeral(s, i, "Talebi üstlendin.")
		_, _ = s.ChannelMessageSend(i.ChannelID, "🙋 Bu talebi <@"+userOf(i).ID+"> üstlendi.")
	case "ticket_add_btn":
		return ticketModal(s, i, "ticket_add_modal", "Kanala Üye Ekle", "user_input", "Üye ID", discordgo.TextInputShort, "")
	case "ticket_rename_btn":
		return ticketModal(s, i, "ticket_rename_modal", "Ticket Kanalını Adlandır", "name_input", "Yeni kanal adı", discordgo.TextInputShort, "")
	case "ticket_status":
		states := []string{"open", "in_progress", "waiting"}
		next := states[0]
		for x, v := range states {
			if t.Status == v {
				next = states[(x+1)%len(states)]
			}
		}
		_, e = b.db.SetTicketStatus(i.ChannelID, next)
		if e != nil {
			return e
		}
		return ephemeral(s, i, "Ticket durumu: **"+next+"**")
	}
	return nil
}
func ticketModal(s *discordgo.Session, i *discordgo.InteractionCreate, id, title, inputID, label string, style discordgo.TextInputStyle, value string) error {
	return s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{Type: discordgo.InteractionResponseModal, Data: &discordgo.InteractionResponseData{CustomID: id, Title: title, Components: []discordgo.MessageComponent{row(discordgo.TextInput{CustomID: inputID, Label: label, Style: style, Required: id != "ticket_close_modal", Value: value, MaxLength: 500})}}})
}
func (b *Bot) ticketAddFromModal(s *discordgo.Session, i *discordgo.InteractionCreate, v map[string]string) error {
	c := &commandContext{b: b, s: s, guildID: i.GuildID, channelID: i.ChannelID, user: userOf(i), member: i.Member}
	u, m, e := b.target(c, v["user_input"])
	if e != nil || m == nil {
		return fmt.Errorf("kullanıcı bulunamadı")
	}
	allow := int64(discordgo.PermissionViewChannel | discordgo.PermissionSendMessages | discordgo.PermissionReadMessageHistory)
	if e = s.ChannelPermissionSet(i.ChannelID, u.ID, discordgo.PermissionOverwriteTypeMember, allow, 0); e != nil {
		return e
	}
	return ephemeral(s, i, "<@"+u.ID+"> ticket kanalına eklendi.")
}
func (b *Bot) ticketRenameFromModal(s *discordgo.Session, i *discordgo.InteractionCreate, v map[string]string) error {
	name := safeChannelName(v["name_input"])
	if len(name) < 2 {
		return fmt.Errorf("geçerli kanal adı belirt")
	}
	if _, e := s.ChannelEdit(i.ChannelID, &discordgo.ChannelEdit{Name: name}); e != nil {
		return e
	}
	return ephemeral(s, i, "Kanal adı **"+name+"** oldu.")
}
func (b *Bot) closeTicketFromModal(s *discordgo.Session, i *discordgo.InteractionCreate, v map[string]string) error {
	t, e := b.db.Ticket(i.ChannelID)
	if e != nil || t.ClosedAt.Valid {
		return fmt.Errorf("bu kanal açık ticket değil")
	}
	c := &commandContext{b: b, s: s, guildID: i.GuildID, channelID: i.ChannelID, user: userOf(i), member: i.Member}
	if t.OwnerID != userOf(i).ID && !b.isSupport(c) {
		return fmt.Errorf("bu talebi kapatma yetkin yok")
	}
	reason := valueOr(v["reason_input"], "Sebep belirtilmedi")
	ok, e := b.db.CloseTicket(i.ChannelID, userOf(i).ID, reason)
	if e != nil || !ok {
		return fmt.Errorf("ticket zaten kapatılmış")
	}
	msgs, _ := s.ChannelMessages(i.ChannelID, 100, "", "", "")
	var lines []string
	for x := len(msgs) - 1; x >= 0; x-- {
		m := msgs[x]
		lines = append(lines, fmt.Sprintf("[%s] %s (%s): %s", m.Timestamp.Format(time.RFC3339), m.Author.Username, m.Author.ID, valueOr(m.Content, "[embed/boş mesaj]")))
	}
	data := []byte(strings.Join(lines, "\n"))
	b.ticketLog(i.GuildID, fmt.Sprintf("🔒 Ticket kapatıldı: **%s** · sahibi: <@%s> · kapatan: <@%s> · sebep: **%s**", i.ChannelID, t.OwnerID, userOf(i).ID, trunc(reason, 200)), &discordgo.File{Name: "ticket-" + i.ChannelID + ".txt", ContentType: "text/plain; charset=utf-8", Reader: bytes.NewReader(data)})
	_ = ephemeral(s, i, "Ticket kapatıldı; kanal 5 saniye içinde silinecek.")
	_, _ = s.ChannelMessageSendEmbed(i.ChannelID, embed("🔒 Ticket Kapatıldı", "**Sebep:** "+trunc(reason, 500), colorDanger))
	time.AfterFunc(5*time.Second, func() { _, _ = s.ChannelDelete(i.ChannelID) })
	return nil
}
func (b *Bot) ticketLog(guild, content string, file *discordgo.File) {
	settings, e := b.db.Settings(guild)
	if e != nil || !settings.TicketLogChannelID.Valid {
		return
	}
	send := &discordgo.MessageSend{Content: safeText(content)}
	if file != nil {
		send.Files = []*discordgo.File{file}
	}
	_, _ = b.dg.ChannelMessageSendComplex(settings.TicketLogChannelID.String, send)
}

var _ database.Ticket
